docs/design/OCI Store Cache Worker.md

```diff
@@ -0,0 +1,191 @@
+    GetSubjectReferenceManifest -->|Store operation| Cache_Hit_1
+    Cache_Hit_1{Cache Hit?}
+    Cache_Hit_1 -->|Hit| GetReferrerArtifact
+    Cache_Hit_1 -->|Miss| ORASFetch1
+    ORASFetch1 -->|Fetch| MemoryBasedStore
+    GetReferrerArtifact --> Cache_Hit_2
+    Cache_Hit_2{Cache Hit?}
+    Cache_Hit_2 -->|Hit| PerformVerification
+    Cache_Hit_2 -->|Miss| ORASFetch2
+    ORASFetch2 -->|Fetch| MemoryBasedStore
+    PerformVerification --> ReturnVerifierResult
+    OCIStoreCache --> CacheProvider
+    CacheProvider --> OCIStoreCache
+
+    %% Highlight subprocesses
+    GetReferrerArtifact:::subProcess
+    ORASFetch1:::subProcess
+    ORASFetch2:::subProcess
+    PerformVerification:::subProcess
+
+    %% Add connections to the cache
+    Cache_Hit_1 -.-> OCIStoreCacheWorker
+    MemoryBasedStore -.-> GetReferrerArtifact
+    Cache_Hit_2 -.-> OCIStoreCacheWorker
+    MemoryBasedStore -.-> PerformVerification
+    MemoryBasedStore -.-> OCIStoreCacheWorker
+    OCIStoreCacheWorker -.-> CacheProvider
+    
+
+    %% Define external components
+    classDef external fill:#f0f0f0,stroke:#333,stroke-width:2px;
+    OCIStoreCacheWorker:::external
+    MemoryBasedStore:::external
+    OCIStoreCache:::external
+    CacheProvider:::external
+```
+
+## Component Description
+
+- Referrer Store Architecture
+
+- Cache Provider: Set up to cache data for `GetSubjectDescriptor`, `GetReferenceManifest`.
+
+- OCI Store(local cache): A content store based on file system with the OCI-Image layout.
+  - Both executor and plugin initiation would `CreateStoresFromConfig`, add cache check can help avoid duplication.
+  - task handler is event driven controlled by cache worker: new enqueue item created, last task finished
+  - a write lock is set in Cache Provider to avoid conflict
+  - a read lock/availability map is set for the writing content to avoid dirty read
+
+- Cache Worker:
+  - Manage memory backed content store
+  - Enqueue task in cache provider to write into OCI Store
+  - Check read lock/availability map in cache provider to avoid dirty read
+
+## Comparing with Other Design
+
+### Single OCIStore VS multi-OCIStores(per verifier)
+
+1. Syncing in between those OCI stores will cause design difficulty.
+2. If if OCI stores are verifer-binded, when the verifier is recycled the OCI stores is gone. Otherwise another resource maintainer(provider) is needed.
+3. Maintaining multi-OCIStores means extra memory cost
+
+### Async VS Sync
+
+1. Sync write to OCIStore leads to racing condition. Using lock would naturally heads to more waiting in line.
+2. When doing sync write, both read write racing condition have be handled in verifier process which increase the complexity. When doing async write, those jobs are delegated to **Cache Provider**.
+
+## Supported Limits and Further Considerations
+
+In Ristretto using scenario, do we support multi-notation verifier, in other words do we have to support cache worker with Ristretto
```

Comment:
In the above diagram, is Ristretto at the `Memory Based Store` level?

Reply:
But in general, I feel in-memory cache is not suitable for storing large data such as blobs.

Reply:
Currently we are having `[]byte` for the return value, which we can make some improvements